com.unity.multiplayer.mlapi/Editor/CodeGen/PostProcessorAssemblyResolver.cs

```diff
@@ -0,0 +1,128 @@
+
+                if (_assemblyCache.TryGetValue(cacheKey, out var result))
+                    return result;
+
+                parameters.AssemblyResolver = this;
+
+                var ms = MemoryStreamFor(fileName);
+
+                var pdb = fileName + ".pdb";
+                if (File.Exists(pdb))
+                    parameters.SymbolStream = MemoryStreamFor(pdb);
+
+                var assemblyDefinition = AssemblyDefinition.ReadAssembly(ms, parameters);
+                _assemblyCache.Add(cacheKey, assemblyDefinition);
+                return assemblyDefinition;
+            }
+        }
+
+        private string FindFile(AssemblyNameReference name)
+        {
+            var fileName = _assemblyReferences.FirstOrDefault(r => Path.GetFileName(r) == name.Name + ".dll");
+            if (fileName != null)
+                return fileName;
+
+            // perhaps the type comes from an exe instead
+            fileName = _assemblyReferences.FirstOrDefault(r => Path.GetFileName(r) == name.Name + ".exe");
+            if (fileName != null)
+                return fileName;
+
+            //Unfortunately the current ICompiledAssembly API only provides direct references.
+            //It is very much possible that a postprocessor ends up investigating a type in a directly
+            //referenced assembly, that contains a field that is not in a directly referenced assembly.
+            //if we don't do anything special for that situation, it will fail to resolve.  We should fix this
+            //in the ILPostProcessing api. As a workaround, we rely on the fact here that the indirect references
+            //are always located next to direct references, so we search in all directories of direct references we
+            //got passed, and if we find the file in there, we resolve to it.
+            foreach (var parentDir in _assemblyReferences.Select(Path.GetDirectoryName).Distinct())
+            {
+                var candidate = Path.Combine(parentDir, name.Name + ".dll");
+                if (File.Exists(candidate))
+                    return candidate;
+            }
+
+            return null;
+        }
+
+        static MemoryStream MemoryStreamFor(string fileName)
+        {
+            return Retry(10, TimeSpan.FromSeconds(1), () =>
+            {
+                byte[] byteArray;
+                using (var fs = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
+                {
+                    byteArray = new byte[fs.Length];
+                    var readLength = fs.Read(byteArray, 0, (int)fs.Length);
+                    if (readLength != fs.Length)
+                        throw new InvalidOperationException("File read length is not full length of file.");
+                }
+
+                return new MemoryStream(byteArray);
+            });
+        }
+
+        private static MemoryStream Retry(int retryCount, TimeSpan waitTime, Func<MemoryStream> func)
```

Comment:
Weird, does this actually happen?  That some file we want to read isn't ready but becomes ready later?

Reply:
ILPPs generally run in parallel and sometimes ILPP'd assembly might be in the "writing to disk" state which wouldn't allow us to read it.